hooks/tk-multi-publish2/basic/collector.py

```diff
@@ -0,0 +1,355 @@
+        """
+        Process the supplied file path.
+
+        :param parent_item: parent item instance
+        :param path: Path to analyze
+        :param frame_sequence: Treat the path as a part of a sequence
+        :returns: The item that was created
+        """
+
+        # make sure the path is normalized. no trailing separator, separators
+        # are appropriate for the current os, no double separators, etc.
+        path = sgtk.util.ShotgunPath.normalize(path)
+
+        publisher = self.parent
+
+        # get info for the extension
+        item_info = self._get_item_info(path)
+        item_type = item_info["item_type"]
+        type_display = item_info["type_display"]
+        evaluated_path = path
+        is_sequence = False
+
+        if frame_sequence:
+            # replace the frame number with frame spec
+            seq_path = publisher.util.get_frame_sequence_path(path)
+            if seq_path:
+                evaluated_path = seq_path
+                type_display = "%s Sequence" % (type_display,)
+                item_type = "%s.%s" % (item_type, "sequence")
+                is_sequence = True
+
+        display_name = publisher.util.get_publish_name(
+            path, sequence=is_sequence)
+
+        # create and populate the item
+        file_item = parent_item.create_item(
+            item_type, type_display, display_name)
+        file_item.set_icon_from_path(item_info["icon_path"])
+
+        # if the supplied path is an image, use the path as # the thumbnail.
+        if (item_type.startswith("file.image") or
+            item_type.startswith("file.texture")):
+            file_item.set_thumbnail_from_path(path)
+
+            # disable thumbnail creation since we get it for free
+            file_item.thumbnail_enabled = False
+
+        # all we know about the file is its path. set the path in its
+        # properties for the plugins to use for processing.
+        file_item.properties["path"] = evaluated_path
+
+        if is_sequence:
+            # include an indicator that this is an image sequence and the known
+            # file that belongs to this sequence
+            file_item.properties["is_sequence"] = True
+            file_item.properties["sequence_files"] = [path]
+
+        self.logger.info("Collected file: %s" % (evaluated_path,))
+
+        return file_item
+
+    def _collect_folder(self, parent_item, folder):
+        """
+        Process the supplied folder path.
+
+        :param parent_item: parent item instance
+        :param folder: Path to analyze
+        :returns: The item that was created
+        """
+
+        # make sure the path is normalized. no trailing separator, separators
+        # are appropriate for the current os, no double separators, etc.
+        folder = sgtk.util.ShotgunPath.normalize(folder)
+
+        publisher = self.parent
+        img_sequences = publisher.util.get_frame_sequences(
```

Comment:
How slow is this on a network fileserver, on Windows, with a somewhat-shitty network? It's just a thought, but Francis can test that with his setup. If it is slow, file IO is something that is very effectively handled by threading in Python since IO doesn't trigger the GIL.